compiler/ir/backend.common/src/org/jetbrains/kotlin/ir/util/IrTypeUtils.kt

```diff
@@ -89,3 +92,38 @@ fun IrType.getPrimitiveArrayElementType() = (this as? IrSimpleType)?.let {
 
 fun IrType.isNonPrimitiveArray() =
     (this.isArray() || this.isNullableArray()) && !this.isPrimitiveArray()
+
+fun IrType.eraseTypeParameters() = when (this) {
+    is IrErrorType -> this
+    is IrSimpleType -> {
+        val owner = classifier.owner
+        when (owner) {
+            is IrClass -> this
+            is IrTypeParameter -> {
+                val upperBound = owner.upperBoundClass()
+                IrSimpleTypeImpl(
+                    upperBound.symbol,
+                    hasQuestionMark,
+                    List(upperBound.typeParameters.size) { IrStarProjectionImpl },
+                    owner.annotations
+                )
+            }
+            else -> error("Unknown IrSimpleType classifier kind: $owner")
+        }
+    }
+    else -> error("Unknown IrType kind: $this")
+}
+
+fun IrTypeParameter.upperBoundClass(): IrClass {
+    // Look into super types for possible upperbound
+    val simpleSuperClassifiers = superTypes.asSequence().filterIsInstance<IrSimpleType>().map { it.classifier }
+    return simpleSuperClassifiers
+        .filterIsInstance<IrClassSymbol>()
+        .let {
+            // First try to find a non-interface class in super types, return it as upperbound.
+            it.firstOrNull { !it.owner.isInterface } ?: it.firstOrNull()
+        }?.owner ?: // Try to find first superType of interface as upperbound
+    // If an upperbound for above condition is not found,
+    // upperbound is in type parameter, dive into type parameters for chained type parameters like U:V, V:String
+    simpleSuperClassifiers.filterIsInstance<IrTypeParameterSymbol>().map { it.owner.upperBoundClass() }.first()
```

Comment:
Instead of invoking `map` on the whole list and taking only the first element from the result, it'd be better to get the first element from the list and run the lambda on it